timely_beliefs/beliefs/utils.py

```diff
@@ -749,3 +751,21 @@ def extreme_timedeltas_not_equal(
     if isinstance(td_a, pd.Timedelta):
         td_a = td_a.to_pytimedelta()
     return td_a != td_b
+
+
+def downsample_first(df: pd.DataFrame, resolution: timedelta) -> pd.DataFrame:
+    """Resample data representing instantaneous events.
+
+    Updates the data frequency, while keeping the event resolution.
+
+    Note that the data frequency may not be constant due to DST transitions
+    The duration between observations is longer for the fall DST transition,
+    and shorter for the spring DST transition.
+    """
+    ds_index = df.index.floor(
+        resolution, ambiguous=[True] * len(df), nonexistent="shift_forward"
```

Comment:
Why can we assume all our datetimes are DST?

Reply:
The `True` value is only used for ambiguous times (see https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DatetimeIndex.floor.html). Imagining that times moves from left to right, this means that for the DST transition in fall we fall back and wait out the 'ambiguous' fold on the left side before continuing to the next event, which creates an extended event. For the DST transition in spring, we spring forward and hop over the 'nonexistent' gap to continue with the next event on the right side, which creates a contracted event.

Sorry for all the terminology. For your convenience:
- fold and gap: used by the [Python datetime module](https://peps.python.org/pep-0495/#terminology)
- fall and spring: names of the seasons in which these DST transitions occur
- fall back and spring forward: a popular mnemonic to remember what happens when
- ambiguous and nonexistent: used by the Pandas package to describe the effects when needing to [localize](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DatetimeIndex.tz_localize.html) timezone naive datetimes around DST transitions (also relevant to the [floor](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DatetimeIndex.floor.html) operation)
- extended event and contracted event: used by timely-beliefs to describe the effect on the resolution of events when resampling events that include a DST transition

Reply:
We do assume that all our datetimes are ambiguous, then? No matter if DST transitions are present or relevant in the data?

Reply:
No, the ambiguous rule is only applies to datetimes that are ambiguous. But nevertheless the array length has to match the frame length.

Reply:
Here, `True` is applied to all datetimes, ambiguous or not.

Reply:
Can you show me what you mean, because I'm printing out the downsampled results and I'm seeing what I expect: that is, mixed offsets, instead of just DST offsets (which is +02 for our tests in the Europe/Amsterdam timezone).

I know that we are passing a `True` value to all datetimes, but pandas is not applying that value to any datetime that is not ambiguous. See the docs for the floor operation, parameter `ambiguous`, third bullet point:
> - bool-ndarray where True signifies a DST time, False designates a non-DST time (note that this flag is only applicable for ambiguous times)

Therefore, I see no point in first determining which **non**-ambiguous datetimes are DST and which are non-DST.

Reply:
That bracketed comment in the third bullet point in another library is easy to overlook. Can you add "pandas is not applying that value to any datetime that is not ambiguous" as a comment here?

Reply:
I also didn't read the Pandas docs like you interpreted it here. But it seems to work.